fairseq/training.py

```diff
@@ -0,0 +1,502 @@
+import logging
+import os
+import subprocess as sp
+import yaml
+import sys
+import copy
+import collections.abc
+import re
+from typing import Dict, Any
+
+from sisyphus import Task, Job, tk
+from sisyphus.hash import sis_hash_helper
+
+import i6_core.util as util
+
+
+class FairseqHydraConfig:
+    """
+    An object that manages a Fairseq hydra config.
+    """
+
+    def __init__(
+        self,
+        config_dict: Dict[str, Any],
+        post_config_dict: Optional[Dict[str, Any]] = None,
+        *,
+        package_name: str = "",
+    ):
+        """
+        :param config_dict: Contains the information which is needed for fairseq-hydra-train. Will be converted and dumped into a .yaml
+        :param dict post_config_dict: dictionary of the FairseqHydraConfig config variables that are not hashed
+        :param package_name: The @package directory that is required to be added to the top of Hydra config, for example "# @package _group_"
+        """
+        assert isinstance(config_dict, dict)
+        assert isinstance(post_config_dict, dict) or post_config_dict is None
+        self.config_dict = config_dict
+        self.post_config_dict = post_config_dict if post_config_dict is not None else {}
+        self.package_name = package_name
+        self.check_consistency()
+
+    @staticmethod
+    def update_nested_dict(dict1, dict2):
+        for k, v in dict2.items():
+            if isinstance(v, collections.abc.Mapping):
+                dict1[k] = update_nested_dict(dict1.get(k, {}), v)
+            else:
+                dict1[k] = v
+        return dict1
+
+    def write(self, path: str):
+        path_corrected_config = self.config_dict.copy()
+        FairseqHydraConfig.update_nested_dict(
```

Comment:
I'd prefer to use the returned value to improve the readability and make clear which variable changes.
```suggestion
        path_corrected_config = FairseqHydraConfig.update_nested_dict(
```